src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+     * Case 2: we have already synced (once or more) during the current
+     * term, but need to sync again (e.g. re-starting the archiver to apply
+     * topic configs)
+     *
+     * They are handled separately to the reader's delight. In theory, it is
+     * possible to fallthrough from Case 1 to Case 2. This function should be
+     * called in a loop until it succeeds.
+     */
+    if (!_raft->is_leader()) {
+        co_return false;
+    }
+
+    const auto current_term = _raft->term();
+    const auto last_applied_offset = _manifest->get_insync_offset();
+
+    if (current_term == _insync_term && _last_replicated_offset) {
+        if (last_applied_offset < _last_replicated_offset) {
+            vlog(
+              _logger.debug,
+              "Syncing archival STM from offset {} to offest in term {}",
+              last_applied_offset,
+              _last_replicated_offset);
+
+            const auto deadline = model::timeout_clock::now() + timeout;
+            const bool synced = co_await wait_no_throw(
+              *_last_replicated_offset, deadline);
+
+            if (!synced) {
+                vlog(
+                  _logger.warn,
+                  "Failed to sync archival STM from offset {} to offest {} in "
+                  "term {}",
+                  last_applied_offset,
+                  _last_replicated_offset,
+                  current_term);
+                co_return false;
+            }
+        }
+
+        if (!_raft->is_leader()) {
+            vlog(
+              _logger.info,
+              "Archival STM sync from offset {} to offest {} in "
+              "term {} was successful, but leadership moved",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term);
+            co_return false;
+        }
+
+        if (current_term != _raft->term()) {
+            vlog(
+              _logger.info,
+              "Archival STM sync from offset {} to offest {} in "
+              "term {} was successful, but term changed to {}",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term,
+              _raft->term());
+            co_return false;
+        }
+
+        if (
+          _last_replicated_offset
+          && _manifest->get_insync_offset() < _last_replicated_offset) {
+            vlog(
+              _logger.warn,
+              "Archival STM sync from offset {} to offest in {}"
+              "term {} was successful, but the last replicated offset moved to "
+              "{}",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term,
+              *_last_replicated_offset);
+            co_return false;
+        }
+
+        co_return true;
+    } else {
+        vlog(
+          _logger.debug,
+          "Syncing archival STM to latest term {}",
+          _raft->term());
+        const bool synced = co_await persisted_stm::sync(timeout);
```

Comment:
nit: i'm wondering if we should call sync in both if/else blocks. Seems like its checks overlap with some in the above case (just the check for _last_replicated_offset is new, no?

Reply:
Yeah. I considered it too, but really there's no guarantee that we'll still be in the same term after the function returns. This means that the archiver needs to check anyways if it is still supposed to operate. So it'll call sync again in that case.